### PR TITLE
[receiver/otlp] Return proper http response code based on retryable errors

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:**
Updates the receiver's http response to return a proper http status based on whether or not the pipeline returned a retryable error.  Builds upon the work done in https://github.com/open-telemetry/opentelemetry-collector/pull/8080 and https://github.com/open-telemetry/opentelemetry-collector/pull/9307

**Link to tracking Issue:**

Closes https://github.com/open-telemetry/opentelemetry-collector/issues/9337
Closes https://github.com/open-telemetry/opentelemetry-collector/issues/8132
Closes https://github.com/open-telemetry/opentelemetry-collector/issues/9636
Closes https://github.com/open-telemetry/opentelemetry-collector/issues/6725

**Testing:**

Updated lots of unit tests

### CODE REVIEW

File: receiver/otlpreceiver/internal/errors/errors.go
Comment (evan-bradley) [inline receiver/otlpreceiver/internal/errors/errors.go:44]: Permanent errors will be [converted to `status.InvalidArgument`](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/internal/errors/errors.go#L20), which is equivalent to an HTTP 400. We should change one of the two codes so they're semantically equivalent between gRPC and HTTP.

Comment (TylerHelmuth) [reply receiver/otlpreceiver/internal/errors/errors.go:44]: I thought about that too, but since [the spec says](https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#failures-1):

```
This specification does not use Status.code field and the server MAY omit Status.code field. The clients are not expected to alter their behavior based on Status.code field but MAY record it for troubleshooting purposes.
```

I opted for now for 500 since it is what we were returning before.  If we want to change something we should change `GetStatusFromError` to return `codes. INTERNAL` or `codes.UNKNOWN`, but we aren't required to by the spec.

Comment (TylerHelmuth) [reply receiver/otlpreceiver/internal/errors/errors.go:45]: With https://github.com/open-telemetry/opentelemetry-collector/pull/9415 merged the status codes are now aligned.

File: receiver/otlpreceiver/otlphttp.go
Comment (bogdandrutu) [inline receiver/otlpreceiver/otlphttp.go:46]: I don't like that now we also have a similar but different logic in `errorMsgToStatus`. Can we consolidate?

Comment (TylerHelmuth) [reply receiver/otlpreceiver/otlphttp.go:46]: Agreed. I took a look at consolidating before opening the PR. I believe it can be refactored, but it caused a lot of changes unrelated to the issue I was trying to solve. To keep the PRs smaller and targeted I'd like to use this PR as a solution to the issue and then a future PR that is only a refactor. I can open an issue to track that refactor if you'd like.

Comment (TylerHelmuth) [reply receiver/otlpreceiver/otlphttp.go:46]: @bogdandrutu are you ok with that approach?

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-maintainers please review

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-maintainers please review

**Comment by mx-psi** [comment]
@bogdandrutu PTAL

**Comment by 0x006EA1E5** [comment]
Hi, I was just looking at this issue myself locally, and have found this open issue...

I notice this, #9307 and #8080 are somewhat coupled to GRCP status codes.

I also notice we have [`consumer/consumererror`](https://github.com/open-telemetry/opentelemetry-collector/tree/main/consumer/consumererror), which seems to address somewhat similar concerns, i.e. retryable and permanent errors.

Can you clarify if `consumer/consumererror` would be appropriate here, or is it intended for something else?  I see `consumererror.New` is [found in a number of places in the contrib codebase](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-collector-contrib%20%22consumererror.New%22&type=code).

Have you considered using `consumererror.IsPermanent(err)` in the receiver to determine the response status code?

Using `consumererror` would seem to make it a bit more user friendly to actually create the errors in the downstream processor, e.g., something like `return td, consumererror.NewTraces(errDataRefused, td)` in [`memorylimiter`](https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/memorylimiterprocessor/memorylimiter.go#L62).

**Comment by TylerHelmuth** [comment]
@0x006EA1E5 as you've deduced, `consumer/consumererror` is normally used downstream from receivers to tell receivers what kind of errors they got from `r.nextConsumer.Consume*`.  In this instance we're in a receiver and the error we're propagating back to the caller is subject to the OTLP specification (which is closely tied to grpc objects even for http) so returning `consumererror` would not be appropriate. [The receiver does use `consumererror` when initially handling the error returned from downstream components.](https://github.com/open-telemetry/opentelemetry-collector/blob/47de8647738d93209537f559874be6c303d3666b/receiver/otlpreceiver/internal/errors/errors.go#L19)

**Comment by TylerHelmuth** [comment]
@bogdandrutu @open-telemetry/collector-approvers please take a look. `errors.GetHTTPStatusCodeFromStatus(err)` does resemble some other internal functions for handling errors in other code paths, but I think a separate refactor after this is added will be best to keep this PR small and targeted.

File: receiver/otlpreceiver/internal/errors/errors.go
Comment (keruitan-wk) [inline receiver/otlpreceiver/internal/errors/errors.go:42]: Is it possible to add a case for returning 429 or a custom HTTP status code that would address https://github.com/open-telemetry/opentelemetry-collector/issues/6725?

Comment (TylerHelmuth) [reply receiver/otlpreceiver/internal/errors/errors.go:42]: @keruitan-wk the [otlp htttp throttling specification](https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#otlphttp-throttling) states:

> If the server receives more requests than the client is allowed or the server is overloaded, the server SHOULD respond with HTTP 429 Too Many Requests or HTTP 503 Service Unavailable and MAY include [“Retry-After”](https://tools.ietf.org/html/rfc7231#section-7.1.3) header with a recommended time interval in seconds to wait before retrying.

My interpretation of that specification is that `503` is an adequate response code to communicate to clients that the request can be retried.  I don't see the spec making any distinction on how a client should handle a `429` vs a `503` After this PR is merged the `otlpreceiver` will be able to return `503`s which will trigger retry attempts from a client. 

I've updated the description to close https://github.com/open-telemetry/opentelemetry-collector/issues/6725 when this PR is merge.

Comment (keruitan-wk) [reply receiver/otlpreceiver/internal/errors/errors.go:42]: HTTP 4xx status codes indicate that it is a client side error, while 5xx status codes indicate server side error. While both 429 and 503 are retryable, it would be nice to be able to distinguish between the two as 429 is meant to tell the client that they are sending too many requests while the collector is working as expected, vs 503 that something is wrong with the collector and they can try again later.

Comment (TylerHelmuth) [reply receiver/otlpreceiver/internal/errors/errors.go:42]: I agree with you, but the client response to either code is the same according to the OTLP specification: exponential backoff retries.

I agree that getting the exactly proper http status code would be nice.  I could add a check in the switch statement for the grpc status code `RESOURCE_EXHAUSTED` which I think is the code that translates to `429`.

**Review by jesusvazquez — APPROVED** [review]
LGTM I've personally tested this change on tempo and has worked as expected, we get proper mappings now. Tempo uses the receiver code for both handling and being up to date with the latest otel changes.

**Review by jpkrohling — APPROVED** [review]

**Review by songy23 — APPROVED** [review]
LGTM

**Review by mx-psi — APPROVED** [review]

File: receiver/otlpreceiver/otlphttp.go
Comment (TylerHelmuth) [reply receiver/otlpreceiver/otlphttp.go:46]: Created https://github.com/open-telemetry/opentelemetry-collector/issues/9864

**Pull request merged**
